### PR TITLE
[BTS-731]

### PR DESCRIPTION
**Pull request by cpjulia**
### Scope & Purpose

This PR disallows index creation for single server or for cluster when the instance is coordinator in a special case. Indexes cannot be created covering fields that have preceding or trailing ":", because the case of preceding or trailing ":" is treated as a special case for shardKeys, in which the value of the attribute is read until or starting from where the character ":" of the string is reached, and it doesn't happen for index fields. We don't disallow this usage for dbservers because this check must be only done for indexes that will be created, not for indexes that already exist. Example: 
for `shardKeys: {"value:"}`, if the document has an attribute `value: "123:abc"`, the shard key would cover "123", which is the substring read until we reach a ":"

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [x] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [x] **integration tests**
  - [ ] **resilience tests**
- [x] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
 

#### Related Information

- [x] Docs PR: https://github.com/arangodb/docs/pull/1086
- [ ] Enterprise PR:
- [x] GitHub issue / Jira ticket: https://arangodb.atlassian.net/browse/BTS-731
- [ ] Design document:

### CODE REVIEW

File: CHANGELOG [hunk 1]
```diff
@@ -1,6 +1,11 @@
 devel
 -----
 
+* Disallowed index creation that covers fields in which the field's name starts
+  or ends with `:` for single server or cluster when the instance is
+  coordinator. This only happens for index creation, so already existing
+  indexes that might have such conditions will remain as they are.
```
Comment (jsteemann) [inline CHANGELOG:7]:
```suggestion
* Disallowed index creation that covers fields in which the field's name starts
  or ends with `:` for single server or cluster when the instance is a
  coordinator or single server. This validation only happens for index creation, 
  so already existing indexes that might use such field names will remain as they are.
```

File: arangod/Indexes/Index.cpp [hunk 2]
```diff
@@ -261,6 +261,20 @@ size_t Index::sortWeight(arangodb::aql::AstNode const* node) {
   }
 }
 
+/// @brief validate fields that start or end with ":"
+void Index::validateFieldsWithSpecialCase(VPackSlice const& fields) {
```
Comment (jsteemann) [inline arangod/Indexes/Index.cpp:265]: Slices are lightweight and only contain a data member, which is a pointer. They can and should be passed by value.
```suggestion
void Index::validateFieldsWithSpecialCase(VPackSlice fields) {
```

File: arangod/Indexes/Index.cpp [hunk 3]
```diff
@@ -261,6 +261,20 @@ size_t Index::sortWeight(arangodb::aql::AstNode const* node) {
...
+  if (!fields.isArray()) {
+    return;
+  }
+  for (VPackSlice name : VPackArrayIterator(fields)) {
+    if (name.toString().starts_with(":") || name.toString().ends_with(":")) {
```
Comment (jsteemann) [inline arangod/Indexes/Index.cpp:270]: Avoid creation of 2 `std::string` temporaries.
```suggestion
    if (name.stringView().starts_with(":") || name.stringView().ends_with(":")) {
```

File: arangod/Indexes/Index.cpp [hunk 4]
```diff
@@ -261,6 +261,20 @@ size_t Index::sortWeight(arangodb::aql::AstNode const* node) {
...
+      THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_ARANGO_ATTRIBUTE_PARSER_FAILED,
+                                     "field names starting with \":\" or "
+                                     "ending with \":\" are disallowed");
```
Comment (jsteemann) [inline arangod/Indexes/Index.cpp:273]:
```suggestion
                                     "field names starting or ending with \":\"are disallowed in indexes");
```

File: arangod/Indexes/Index.h [hunk 5]
```diff
@@ -258,6 +258,9 @@ class Index {
   /// @brief whether or not the index is unique
   inline bool unique() const { return _unique; }
 
+  /// @brief validate fields that start or end with ":"
+  static void validateFieldsWithSpecialCase(velocypack::Slice const& fields);
```
Comment (jsteemann) [inline arangod/Indexes/Index.h:262]:
```suggestion
  /// @brief validates that field names don't start or end with ":"
  static void validateFieldsWithSpecialCase(velocypack::Slice fields);
```

File: arangod/VocBase/Methods/Indexes.cpp [hunk 6]
```diff
@@ -397,6 +397,21 @@ Result Indexes::ensureIndex(LogicalCollection* collection, VPackSlice input,
...
   VPackSlice indexDef = normalized.slice();
+  // for single server or for cluster when the instance is coordinator,
+  // indexes cannot be created covering fields that have preceding or trailing
+  // ":", because the case of preceding or trailing ":" is treated as a special
+  // case for shardKeys, in which the value of the attribute is read until or
+  // starting from where the character ":" of the string is reached, and it
+  // doesn't happen for index fields. We don't disallow this usage for dbservers
+  // because this check must be only done for indexes that will be created, not
+  // for indexes that already exist. Example for shardKeys: {"value:"}, if the
+  // document has an attribute "value: 123:abc", the shard key would cover
```
Comment (jsteemann) [inline arangod/VocBase/Methods/Indexes.cpp:408]:
```suggestion
  // document has an attribute {"value": "123:abc"}, the shard key would cover
```

File: tests/js/common/shell/shell-index.js [hunk 7]
```diff
@@ -266,7 +266,32 @@ function IndexSuite() {
...
       }
-    }
+    },
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief create index with field names that start or end with ":"
+////////////////////////////////////////////////////////////////////////////////
+
+    testCreateInvalidField: function() {
+      const indexTypes = ["geo", "fulltext", "persistent", "hash"];
```
Comment (jsteemann) [inline tests/js/common/shell/shell-index.js:276]: "hash" is an alias type for a "persistent" index. No need to test it separately.
```suggestion
      const indexTypes = ["geo", "fulltext", "persistent"];
```

File: arangod/Indexes/Index.cpp [hunk 8]
```diff
@@ -261,6 +261,21 @@ size_t Index::sortWeight(arangodb::aql::AstNode const* node) {
...
+void Index::validateFieldsWithSpecialCase(VPackSlice fields) {
+  if (!fields.isArray()) {
+    return;
+  }
+  for (VPackSlice name : VPackArrayIterator(fields)) {
+    if (name.stringView().starts_with(":") ||
+        name.stringView().ends_with(":")) {
+      THROW_ARANGO_EXCEPTION_MESSAGE(
+          TRI_ERROR_ARANGO_ATTRIBUTE_PARSER_FAILED,
+          "field names starting or ending with \":\"are disallowed in indexes");
```
Comment (jsteemann) [inline arangod/Indexes/Index.cpp:274]:
```suggestion
          "field names starting or ending with \":\" are disallowed in indexes");
```

File: arangod/Indexes/Index.cpp [hunk 9]
```diff
@@ -261,6 +261,25 @@ size_t Index::sortWeight(arangodb::aql::AstNode const* node) {
...
+
+  for (VPackSlice name : VPackArrayIterator(fields)) {
+    if (name.isObject()) {
+      name = name.get("name");
+    }
+    if (name.stringView().starts_with(":") ||
+        name.stringView().ends_with(":")) {
+      THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_ARANGO_ATTRIBUTE_PARSER_FAILED,
+                                     "field names starting or ending with "
+                                     "\":\"are disallowed in indexes");
```
Comment (jsteemann) [inline arangod/Indexes/Index.cpp:278]:
```suggestion
                                     "\":\" are disallowed in indexes");
```

**Review by jsteemann — APPROVED** [review]
LGTM

File: arangod/VocBase/Methods/Indexes.cpp [hunk 10]
```diff
@@ -397,6 +397,21 @@ Result Indexes::ensureIndex(LogicalCollection* collection, VPackSlice input,
...
 
   VPackSlice indexDef = normalized.slice();
+  // for single server or for cluster when the instance is coordinator,
+  // indexes cannot be created covering fields that have preceding or trailing
+  // ":", because the case of preceding or trailing ":" is treated as a special
+  // case for shardKeys, in which the value of the attribute is read until or
+  // starting from where the character ":" of the string is reached, and it
+  // doesn't happen for index fields. We don't disallow this usage for dbservers
+  // because this check must be only done for indexes that will be created, not
+  // for indexes that already exist. Example for shardKeys: {"value:"}, if the
```
Comment (Simran-B) [inline arangod/VocBase/Methods/Indexes.cpp:407]:
```suggestion
  // for indexes that already exist. Example for shardKeys: ["value:"], if the
```

**Review by Simran-B — CHANGES_REQUESTED** [review]

**Pull request merged**